### ISSUE TITLE
Used variable is being reported as unused in a small edge case.

### ISSUE THREAD

**Issue by captbaritone** [eslint/eslint#6703]
**What version of ESLint are you using?**

3.1.1

**What parser (default, Babel-ESLint, etc.) are you using?**
Online Demo

**Please show your full configuration:**

```js
{
    "rules": {
        "no-unused-vars": "error"
    }
}
```

**What did you do? Please include the actual source code causing the issue.**

```js
function logDigits(string) {
    string = string.replace(/\d/, console.log)
}

logDigits('abc-123');
```

**What did you expect to happen?**

I expected to see an error on line 2 indicating that I didn't use the "replaced" version of `string`.

**What actually happened? Please include the actual, raw output from ESLint.**

```sh
  1:20  error  'string' is defined but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)
```

This code does have an issue, the assigned version of `string` is not used and should be removed. However, the way it is reported seems to indicate that the string argument is not used at all, which is not true since the execution of the `replace` callback is dependant upon which the value of `string` that you pass it.

I'm guessing this is related to https://github.com/eslint/eslint/issues/6348.

**Comment by platinumazure** [comment]
Seems to me like you would just want to invoke `string.replace` for its side effects:

```js
function logDigits(string) {
    string.replace(/\d/, console.log);
}

logDigits('abc-123');
```

Am I missing something here?

**Comment by captbaritone** [comment]
@platinumazure Correct. Removing the assignment is the correct fix to the code, and my concern it is showing a misleading message on a misleading line.

The self-assignment check seems to be trumping the fact that I _am_ using `string` inside this function.

**Comment by platinumazure** [comment]
Yeah, I guess I'm not sure why that was implemented as default behavior instead of an option. My vote was to have an option.

If we are going to leave this as default behavior, then I would say we should change the message for the self-assign case. Maybe something like `"Self-assigned variable 'string' is not used anywhere else"`.

**Comment by mikesherov** [comment]
I think it's a bug fix that breaks builds and so was therefore minor (not patch).

**Comment by captbaritone** [comment]
If we change the error message to "Self-assigned variable 'string' is not used anywhere else" we would also need to move it to the line where the self-assignment happens. Currently it is attached to the first instance of the variable: the function argument.

**Comment by ilyavolodin** [comment]
I'm not 100% sure I'm following the argument presented here. In the end, `string` is unused. Whether it was self assigned or not, it's still unused. Consider the following simplified example:
```js
function logDigits(string) {
    string = string
}

logDigits('abc-123');
```
Should `string` not be marked as unused in this case? I'm also not sold on changing the message. Again, since self-assignment is not usage of the variable.

**Comment by captbaritone** [comment]
@ilyavolodin `string` is "used" in so far as changing the value of `string` that you pass the function will change what code gets executed inside the `replace` callback function. In my attempt to make my example minimal, I think I wrote a very confusing example. Perhaps this is a better example:

```
function logDigits(string) {
    var numbers = [];
    string = string.replace(/\d/g, function(match) {
      numbers.push(match);
    });
    return numbers;
}

console.log('Example #1', logDigits('abc-123'));
// => "Example #1", ["1", "2", "3"]

console.log('Example #2', logDigits('4a5a6'));
// => "Example #1", ["4", "5", "6"]
```

https://jsbin.com/nemidi/edit?js,console

As you can see, the `string` argument affects the output of the function, and is therefore "used".

Sorry for the confusing example.

**Comment by ilyavolodin** [comment]
Ah, OK. This example makes much more sense. However, in this case the output of the replace function is never used. And this rule reports the error correctly. The error highlight variable declaration location, which I also think is correct. I'm not sure we can do anything else in this case.

**Comment by captbaritone** [comment]
It's true that the output of `string.replace()` is never used. But in this case the error is reported on line one where the `string` argument of `logDigits` is specified.

**Comment by ilyavolodin** [comment]
Correct. We always report location of variable declaration. In this case, variable is declared as an argument of a function. If we were to switch to the second line, we wouldn't be able to handle a case like this:

```js
function foo(bar) {
  bar = bar;
  bar = bar;
  bar = bar;
}
````

Either we would have to report 3 errors, or we report the location of the error the same way we do now, which is line 1.

**Comment by captbaritone** [comment]
Got it. This is enough of an edge case, that it may not be worth discussing any further, but I wonder if this particular version of this error (unused __re__assignments) should be either a separate rule or trigger a different message.

In this case "is defined but never used" is not really true. The defined variable __is__ used. It's the reassignment that is not used. I would expect an error message like "this assignment of x is never used" or "variable x is reassigned but never used" and I would expect to see it on the actual assignment, not the declaration.

In the chained example you gave, I would expect to see the error on the final assignment. Fixing that error would uncover an identical error on the preceding assignment.

**Comment by ilyavolodin** [comment]
The reason for reporting variable declaration is that in most cases, if the variable isn't used, it's safe to remove it. And to remove it, it's better to know where it's declared. Once you remove declaration, all the uses will be highlighted by `no-undef` rule.

Regarding error message, I don't have a strong feelings about it. In my mind, a variable is a variable, in the end, it's not used, and that's what being reported.

**Comment by platinumazure** [comment]
Since we iterate over all the reads of a variable and we are already discriminating between self-assignment and other uses (and only counting the latter as uses per the rule), I personally think it would be nice to track whether one or more self-assigns have occurred and change the message (without changing the report location) to note that self-assign doesn't count as use.

Or, even better, we could revert the behavior and put it behind an option, so it's at least less surprising to end users.

**Comment by ilyavolodin** [comment]
@platinumazure I disagree, I don't think this should be an option. This is what rule was supposed to catch from the begging, it was just an oversight. As I said, the variable is unused, everything else is just details.

**Comment by mikesherov** [comment]
Agreed, this is why we have our semver policy. Minor can be used for bugfixes that fail builds. :-1: on reverting and :-1: on an option.

**Comment by captbaritone** [comment]
For the purposes of this discussion it might be worth drawing a distinction between a "variable", a "variable declaration" and a "variable assignment".

In this case, it seems to me that the following is true:

| Term | Used | Unused | Rational |
| --- | --- | --- | --- |
| "Variable" | ✔️ | | The `string` variable is used by the `.replace()` method's callback | 
| "Variable Declaration" | ✔️ | | The variable declared as `string` is used by the `.replace()` method's callback |
| "Variable Assignment" |  | ✔️ | The value assinged to `string` by the `.replace()` method is not used |
If others see it differently, that may be the source of some of this disagreement and coming to an agreement on the words we are using may help us avoid confusion and come to a mutual agreement more easily.

**Comment by mysticatea** [comment]
Granted, the error message is confusing in this case.

I think there are 2 solutions:

- **A:** we change the message and the location for unused reassignments.
- **B:** we separate the check of unused reassignments from this rule.

### **A:** we change the message and the location for unused reassignments.

It will move the error location to the last self-reassignment.

```js
function foo(a) {    // 'a' is defined but never used. (no-unused-vars)
    doSomething();
}

function foo(a) {
    doSomething();
    a = a.call();    // 'a' is reassigned but never used. (no-unused-vars)
}

function foo(a) {
    doSomething();
    a = a.call();
    a = a.call();    // 'a' is reassigned but never used. (no-unused-vars)
}

function foo(a) {
    doSomething(a);
    a = a.call();
}
```

### **B:** we separate the check of unused reassignments from this rule.

It will revert the exception of self-assignments, then will add the new rule to warn unused reassignments, probably `no-unused-reassign`.

```js
function foo(a) {    // 'a' is defined but never used. (no-unused-vars)
    doSomething();
}

function foo(a) {
    doSomething();
    a = a.call();    // 'a' is reassigned but never used. (no-unused-reassign)
}

function foo(a) {
    doSomething();
    a = a.call();
    a = a.call();    // 'a' is reassigned but never used. (no-unused-reassign)
}

// one of pros is no-unused-reassign rule can warn for used variables.
function foo(a) {
    doSomething(a);
    a = a.call();    // 'a' is reassigned but never used. (no-unused-reassign)
}
```

----

Thoughts?

**Comment by mikesherov** [comment]
I'm against a new rule for this. Unused is unused whether you've mistakenly re-assigned or not. As @ilyavolodin said, this is how this should've always worked.

Regarding the messaging, our two choices are: keep as is and singularly report where the binding was introduced, or to report ALL the instances of reassignment without use. e.g.:

```
let a;
a = 2;
a = a + 1;
```
either has to report line 1, or report line 1, 2, and 3. I personally think it should report line 1 for the reasons Ilya has already outlined, but also can see the argument for 1, 2, and 3. 

I don't, however, see an argument for reporting only line 3. As a dev, ill delete line 3 naively, re-run ESLint, see now there's the *same* error but on line 2. Frustrated, I then delete line 2, and rerun. Same error again :-(, but now it's on line 1. Even more frustrated, I delete line 1 and re-run. Finally, ESLint passes.

**Comment by BobChao87** [comment]
The problem with reporting on only line 1 in the example proposed by @mikesherov, is that a developer might also go through the same logic as when reporting the error on only the last line. "Oh, this variable is never used, I can safely delete this declaration." And when one does so, then the `no-undef` rule fires and this is bad code. There needs to be some sort of differentiation with the message at a minimum.

Furthermore, according to the rules listed on the `no-unused-vars` docs, this example is a perfect case of a used variable. To quote the docs: "A variable is <i>not</i> considered to be used if it is only ever assigned to (`var x = 5`) or declared." And the more verbose list of what makes a variable used unequivocally includes this use case.

We could potentially address this by altering the docs, but a naïve user will not read this rule's reports how it functions currently. It's easy to tell what's going if the rule suddenly fires in an IDE with ESLint integration as one is coding and one is informed of how this rule works. But from just an output file from the project as a whole, this rule is currently inscrutable as it refers to two different potential problems (with phrasing producing a bias towards believing one of them).

**Comment by nzakas** [comment]
I'm of the mind that we should just update the docs to explain this case. The rule is behaving as intended and the only issue is that some may find the error message confusing. Error messages are short and are not meant to encompass all of the meaning on their own, that is precisely why we have documentation for rules: so we can provide more information than that single phrase can. 

So I'm 👎 for any code changes and 👍 for clarifying this use case in the docs.

**Comment by SpadeAceman** [comment]
I've been following the progression of recent changes to is rule (such as in issues #6348, #2342 and #6646), and while I agree that the rule's current behavior is generally good, an improved error message does seem warranted, based on these recent discussions. 

For what it's worth, here's my suggestion:

**`assignments to variable 'string' are not used elsewhere`**

This would more clearly explain what's being flagged by this rule, encompassing all potential use cases discussed here (from single assignment to multiple self-assignments), while still remaining clear and concise. This should also help retain the current behavior of flagging the variable in question only once.

**Comment by nzakas** [comment]
Changing this over to a documentation task.

**Comment by not-an-aardvark** [comment]
Closing this issue as fixed; the error message was improved in https://github.com/eslint/eslint/pull/7315, which hopefully clarifies the error.
